cmd/gateway-manta.go

```diff
@@ -0,0 +1,368 @@
+	return nil
+}
+
+// StorageInfo - Not relevant to Triton backend.
+func (t *tritonObjects) StorageInfo() (si StorageInfo) {
+	return si
+}
+
+//
+// ~~~ Buckets ~~~
+//
+
+// MakeBucketWithLocation - Create a new directory within manta.
+//
+// https://apidocs.joyent.com/manta/api.html#PutDirectory
+func (t *tritonObjects) MakeBucketWithLocation(bucket, location string) error {
+	ctx := context.Background()
+	err := t.client.Dir().Put(ctx, &storage.PutDirectoryInput{
+		DirectoryName: path.Join(mantaDefaultRootStore, bucket),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetBucketInfo - Get directory metadata..
+//
+// https://apidocs.joyent.com/manta/api.html#GetObject
+func (t *tritonObjects) GetBucketInfo(bucket string) (bi BucketInfo, e error) {
+	var info BucketInfo
+	ctx := context.Background()
+	resp, err := t.client.Objects().Get(ctx, &storage.GetObjectInput{
+		ObjectPath: path.Join(mantaDefaultRootStore, bucket),
+	})
+	if err != nil {
+		return info, err
+	}
+
+	info = BucketInfo{
+		Name:    bucket,
+		Created: resp.LastModified,
+	}
+
+	return info, nil
+}
+
+// ListBuckets - Lists all Manta directories, uses Manta equivalent
+// ListDirectories.
+//
+// https://apidocs.joyent.com/manta/api.html#ListDirectory
+func (t *tritonObjects) ListBuckets() (buckets []BucketInfo, err error) {
+	ctx := context.Background()
+	dirs, err := t.client.Dir().List(ctx, &storage.ListDirectoryInput{
+		DirectoryName: path.Join(mantaDefaultRootStore),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var bucketList []BucketInfo
+	for _, dir := range dirs.Entries {
+		if dir.Type == "directory" {
+			bucket := BucketInfo{
+				Name:    dir.Name,
+				Created: dir.ModifiedTime,
+			}
+
+			bucketList = append(bucketList, bucket)
+		}
+	}
+
+	return bucketList, nil
+}
+
+// DeleteBucket - Delete a directory in Manta, uses Manta equivalent
+// DeleteDirectory.
+//
+// https://apidocs.joyent.com/manta/api.html#DeleteDirectory
+func (t *tritonObjects) DeleteBucket(bucket string) error {
+	ctx := context.Background()
+	return t.client.Dir().Delete(ctx, &storage.DeleteDirectoryInput{
+		DirectoryName: path.Join(mantaDefaultRootStore, bucket),
+	})
+}
+
+//
+// ~~~ Objects ~~~
+//
+
+// ListObjects - Lists all objects in Manta with a container filtered by prefix
+// and marker, uses Manta equivalent ListDirectory.
+//
+// https://apidocs.joyent.com/manta/api.html#ListDirectory
+func (t *tritonObjects) ListObjects(bucket, prefix, marker, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
+	ctx := context.Background()
+	objs, err := t.client.Dir().List(ctx, &storage.ListDirectoryInput{
+		DirectoryName: path.Join(mantaDefaultRootStore, bucket, prefix),
+	})
+	if err != nil {
+		return result, err
+	}
+
+	var blObjects []ObjectInfo
+	var blPrefixes []string
+
+	for _, obj := range objs.Entries {
+
+		if obj.Type == "directory" {
+			blPrefixes = append(blPrefixes, fmt.Sprintf("%s/", obj.Name))
```

Comment:
Is adding prefixes necessary? are you sure that obj.Name doesn't have a "/" at the end?